tests/test_utils.py

```diff
@@ -68,3 +69,35 @@ def test_hms_to_seconds(seconds: str, expected: int):
     result = hms_to_seconds(seconds)
 
     assert result == expected
+
+
+def test_url_connection_raises_on_invalid_url():
+    with pytest.raises(ValueError, match=r".*invalidhosturl.*"):
+        check_url_connection("foo-invalidhosturl")
+
+
+def test_url_connection_checks_hostname(mocker):
+    socket_library_mock = mocker.patch("bonfire.utils.socket.socket")
+    socket_mock = socket_library_mock.return_value.__enter__.return_value
+    check_url_connection("https://validhost.com")
+    socket_mock.connect.assert_called_with(("validhost.com", 443))
+    socket_mock.settimeout.assert_called_once()
+
+
+def test_url_connection_timeout_handling(mocker):
+    socket_library_mock = mocker.patch("bonfire.utils.socket.socket")
+    socket_mock = socket_library_mock.return_value.__enter__.return_value
+    socket_mock.connect.side_effect = TimeoutError("timed out!")
+
+    with pytest.raises(FatalError, match=r"Unable to connect to.*after.*seconds.*is VPN needed.*"):
+        check_url_connection("https://timingout.com")
+
+
+def test_ip_timeout():
+    with pytest.raises(FatalError, match="Unable to connect to.*after 1 seconds.*is VPN needed.*"):
+        check_url_connection("https://10.255.255.1", timeout=1)
```

Comment:
I'm not sure about this, the moment that IP responds back the test is going to fail - however I don't see how "easily" we could mock that. Besides it shouldn't be necessary to actually run a network request externally against the default gateway IP (I think this is what you were trying). Main reason is, if this ever changes, the exception captured in the test is going to fail with a ConnectionRefused - Which - will be shadowed by the OSError and still make the test to pass.

I must say I tried creating a "small" socket server in a fixture so we didn't mock but actually had a socket listening to force the timeout with a custom handler, but that didn't go well. I was never able to make it work (and even if I did, it was going to require either threading or using asyncio, which was building up to be a "way to complex" for the task at hand so it didn't make sense even if I pulled it off)

Anyways - I think I just wanted to make my point, we can leave it as it is - but let me know your thoughts please.

Reply:
Ah - okay, found the [RFC](https://datatracker.ietf.org/doc/html/rfc1918#section-3) - thank you